### PR TITLE
loader/svg: Improve valid check for url(#id)

### PR DESCRIPTION
**Pull request by JSUYA**
Improve parenthesis checking and space checking.
 - There must be only one pair of parentheses.
 - There cannot be spaces(and ') between id strings.

issue: https://github.com/thorvg/thorvg/issues/1983

### CODE REVIEW

**Comment by JSUYA** [comment]
+) Filling the invalid gradient id with black should be considered separately.

**Comment by capnm** [comment]
> +) Filling the invalid gradient id with black should be considered separately.

I compiled  godot with the 2 id PR's, both gradients are now painted black:

image

<details><summary>Test SVG</summary>

```xml
<svg viewBox="0 0 160 160" xmlns="http://www.w3.org/2000/svg">
<defs>
  <linearGradient id="" x2="1">
    <stop stop-color="#0f0"/>
  </linearGradient>
  <linearGradient id="b" x2="1">
    <stop stop-color="#00f"/>
  </linearGradient>
  <linearGradient id="bl ack" x2="1">
    <stop stop-color="#ff0"/>
  </linearGradient>
</defs>

<rect width="160" height="160" fill="red"/>
<rect width="140" height="140" stroke="#f0f" stroke-width="50" fill="url(#)"/>
<rect width="40" height="100" stroke="#0ff" stroke-width="10" fill="url(#bl ack)"/>

</svg>
```
test6


</details>

**Comment by MewPurPur** [comment]
From what I see, it's like:

- url(#) is valid syntax, but references an empty ID, and empty IDs are invalid, therefore should be transparent.

- url(#bl ack) is invalid syntax, therefore it's dismissed, and the attribute it's in should be its default value (i.e. fill should be black). There's no black gradients, just the attribute value you're giving should be dismissed.

**Comment by capnm** [comment]
The current modification doesn't work as intended:
```xml
<svg viewBox="0 0 160 160" xmlns="http://www.w3.org/2000/svg">
<defs>
  <linearGradient id="a" x2="1">
    <stop stop-color="#0f0"/>
    <stop offset="100%" stop-color="blue" />
  </linearGradient>
  <linearGradient id="b" x2="1">
    <stop stop-color="#00f"/>
  </linearGradient>
  <linearGradient id="c" x2="1">
    <stop stop-color="#ff0"/>
    <stop offset="100%" stop-color="blue" />
  </linearGradient>
</defs>

<rect width="160" height="160" fill="url(#a)"/>
<rect width="140" height="140" stroke="#f0f" stroke-width="20" fill="url(#b)"/>
<rect width="130" height="80"  stroke="#0ff" stroke-width="12" fill="url(#c)"/>

</svg>
```

?
    const char* start = hash + 1;
    const char* end = close - 1;
    if (end <= start) return nullptr;

<details><summary>Test</summary>

Expected:
image

PR → svg2png:
test7wip


</details>

**Comment by JSUYA** [comment]
@MewPurPur 
> url(#) is valid syntax, but references an empty ID, and empty IDs are invalid, therefore should be transparent.

I modified to find empty cases. At least in some test  it seems to work fine. Thank you for your review.

@capnm 
Oh my mistake. Thank you for your review. :)

**Comment by mgrudzinska** [comment]
@JSUYA can you please comment on the expected result of the sample?

```
<svg viewBox="0 0 160 160" xmlns="http://www.w3.org/2000/svg">
  <defs>
      <linearGradient id="b " x2="1">
      <stop stop-color="#f0f"/>
    </linearGradient>
  </defs>

  <rect x="30" y="30" width="30" height="30" stroke="#0f0" stroke-width="1" fill="url(#b)"/>
  <rect x="50" y="50" width="30" height="30" stroke="#0f0" stroke-width="2" fill="url(#b )"/>
</svg>
```

from this pr we get:
Zrzut ekranu 2024-02-21 o 01 33 17

from other viewers I got:
Zrzut ekranu 2024-02-21 o 01 35 55

**Comment by MewPurPur** [comment]
IDs are not allowed to have whitespace.

File: src/loaders/svg/tvgSvgLoader.cpp [hunk 1]
```diff
@@ -379,18 +379,29 @@ static void _parseDashArray(SvgLoaderData* loader, const char *str, SvgDash* das
 static char* _idFromUrl(const char* url)
 {
     url = _skipSpace(url, nullptr);
```
Comment (hermet) [inline src/loaders/svg/tvgSvgLoader.cpp:381]: --

File: src/loaders/svg/tvgSvgLoader.cpp [hunk 2]
```diff
@@ -379,18 +379,29 @@ static void _parseDashArray(SvgLoaderData* loader, const char *str, SvgDash* das
...
-        url = _skipSpace(url, nullptr);
-    }
 
-    if ((*url) == '\'') ++url;
-    if ((*url) == '#') ++url;
+    const char* open = strchr(url, '(');
+    const char* close = strchr(url, ')');
+
+    if (!open || !close || close < open) return nullptr;
+    if (strchr(open + 1, '(')) return nullptr;
```
Comment (hermet) [inline src/loaders/svg/tvgSvgLoader.cpp:387]: 387, 388 are not unnecessary compares:

File: src/loaders/svg/tvgSvgLoader.cpp [hunk 3]
```diff
@@ -379,18 +379,29 @@ static void _parseDashArray(SvgLoaderData* loader, const char *str, SvgDash* das
...
+    if (strchr(close + 1, ')')) return nullptr;
+
+    const char* hash = strchr(open, '#');
+    if (!hash || hash > close) return nullptr;
+
+    const char* start = hash + 1;
+    const char* end = close - 1;
+    start = _skipSpace(start, nullptr);
```
Comment (hermet) [inline src/loaders/svg/tvgSvgLoader.cpp:395]: look unnecessary..

**Review by hermet — COMMENTED** [review]
the optimal code we need here:

```
    auto open = strchr(url, '(');
    auto close = strchr(url, ')');
    if (!open || !close || open >= close) return nullptr;

    open = strchr(url, '#');
    if (!open || open >= close) return nullptr;

    ++open;
    --close;

    //trim the rest of the spaces if any
    while (open < close && *close == ' ') --close;

    //quick verification
    for (auto id = open; id < close; id++) {
        if (*id == ' ' || *id == '\'') return nullptr;
    }

    return strDuplicate(open, (close - open + 1));
```

**Comment by hermet** [comment]
for a quick fix and release, I will proceed with the apply regardless of @JSUYA's response.

**Comment by hermet** [comment]
https://github.com/thorvg/thorvg/pull/1984

**Pull request closed without merging**